include/crab/domains/symbolic_variable_eq_domain.hpp

```diff
@@ -0,0 +1,1031 @@
+    }
+
+    m_parents.insert({v, v});
+    m_classes.insert({v, equivalence_class_t(val)});
+  }
+
+  void print_elems_vector(crab::crab_os &o,
+                          const std::vector<element_t> &elems) const {
+    o << "[";
+    for (auto it = elems.begin(), et = elems.end(); it != et;) {
+      o << *it;
+      ++it;
+      if (it != et) {
+        o << ",";
+      }
+    }
+    o << "]";
+  }
+
+  void print_equiv_classes(crab_os &o,
+                           const equivalence_class_elems_t &equiv_classes,
+                           bool verbose = false) const {
+    o << "{";
+    for (auto it = equiv_classes.begin(), et = equiv_classes.end(); it != et;) {
+      print_elems_vector(o, it->second);
+      if (!verbose) {
+        o << "=>" << *(m_classes.at(it->first).get_absval());
+      }
+      ++it;
+      if (it != et) {
+        o << ",";
+      }
+    }
+    o << "}";
+  }
+
+  void print_classes_vals(crab_os &o) const {
+    o << "{";
+    for (auto it = m_classes.begin(), et = m_classes.end(); it != et;) {
+      o << it->first << "=>" << *(it->second.get_absval());
+      ++it;
+      if (it != et) {
+        o << ",";
+      }
+    }
+    o << "}";
+  }
+
+public:
+  /**------------------ Begin union find APIs ------------------**/
+  /// @brief Check whether domain constains element v in some class
+  /// @param v an element
+  /// @return true if the elment exists; otherwise, false.
+  bool contains(const element_t &v) const {
+    return m_parents.find(v) != m_parents.end();
+  }
+
+  /// @brief Check whether two elements in the same class even if they may not
+  /// exist
+  /// @param x an element
+  /// @param y an element differs from x
+  /// @return true if they are in the same class; otherwise, return false
+  bool equals(const element_t &x, const element_t &y) const {
+    if (!contains(x) || !contains(y)) {
+      return false;
+    }
+    const element_t &rep_x = find(x);
+    const element_t &rep_y = find(y);
+    return rep_x == rep_y;
+  }
+
+  /// @brief find the representative without path-compression
+  /// @param v an element in some set
+  /// @attention v must constain in current domain; otherwise, an error returns
+  /// @return returns the representative of the set that contains the element v
+  element_t find(const element_t &v) const {
+    SVEQ_DOMAIN_SCOPED_STATS(".find");
+
+    auto it = m_parents.find(v);
+    if (it == m_parents.end()) {
+      CRAB_ERROR(domain_name(), "::", __func__, " on a non-existing elem ", v,
+                 " in ", *this);
+    }
+    return it->second;
+  }
+
+  boost::optional<element_t> find_opt(const element_t &v) const {
+    SVEQ_DOMAIN_SCOPED_STATS(".find");
+
+    auto it = m_parents.find(v);
+    if (it == m_parents.end()) {
+      return boost::none;
+    }
+    return it->second;
+  }
+
+  /// @brief union two sets
+  /// @param x one set in domain
+  /// @param y one set in domain
+  void union_sets(const element_t &x, const element_t &y) {
```

Comment:
This doesn't seem to be used but it looks like `union` operation. Use this instead of overwritting values in `m_parents` directly (see the example in `add`). Try to find all the places where you should be using `union` (`meet` probably)